8362117: C2: compiler/stringopts/TestStackedConcatsAppendUncommonTrap.java fails with a wrong result due to invalidated liveness assumptions for data phis

danielogh opened the pull request:
This PR addresses a wrong compilation during string optimizations.

During stacked string concatenation of two StringBuilder links SB1 and SB2, the pattern "append -> Phi -> Region -> (True, False) -> If -> Bool -> CmpP -> Proj (Result) -> toString" may be observed, where toString is the end of SB1, and the simple diamond is part of SB2.

After JDK-8291775, the Bool test to the diamond If is set to a constant zero to allow for folding the simple diamond away during IGVN, while not letting the top() value from the result projection of SB1 propagate through the graph too quickly. The assumption was that any data Phi of the Region would go away during PhaseRemoveUseless as they are no longer live -- I think that in the case of JDK-8291775, the user of phi was the constructor of SB2. However, in the attached test case, the Phi stays live as it's a parameter (input to an append) of SB2 and will be used during the transformation in `copy_string`. When the diamond region is later folded, the Phi's user picks up the wrong input corresponding to the false branch.

The proposed solution is to disable the stacked concatenation optimization for this specific pattern. This might be pragmatic as it's an edge case and there's already a bug tail: JDK-8271341-> JDK-8291775 -> JDK-8362117.

Testing: T1-3 (aed5952).

Extra testing: ran T1-3 on Linux with an instrumented build and verified that the pattern I am excluding in this PR is not seen during any other compilation than that of the proposed regression test.

---------
- [ ] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[ \] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Error
&nbsp;⚠️ Pull request body is missing required line: `- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).`

### Issue
 * [JDK-8362117](https://bugs.openjdk.org/browse/JDK-8362117): C2: compiler/stringopts/TestStackedConcatsAppendUncommonTrap.java fails with a wrong result due to invalidated liveness assumptions for data phis (**Bug** - P3)


### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**) 🔄 Re-review required (review applies to [5638f221](https://git.openjdk.org/jdk/pull/27028/files/5638f221ce30776290aa6578fb1200d0a085b623))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/27028/head:pull/27028` \
`$ git checkout pull/27028`

Update a local copy of the PR: \
`$ git checkout pull/27028` \
`$ git pull https://git.openjdk.org/jdk.git pull/27028/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 27028`

View PR using the GUI difftool: \
`$ git pr show -t 27028`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/27028.diff">https://git.openjdk.org/jdk/pull/27028.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/27028#issuecomment-3241125482)
</details>

danielogh committed:
fix

danielogh committed:
revert to unfolded version of is_diamond

danielogh committed:
add an assert

danielogh committed:
ws

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -1062,6 +1067,29 @@ bool StringConcat::validate_control_flow() {
         // transforms the Region. Make sure it sees it.
+
+        // First exclude the following pattern:
+        // append -> Phi -> Region -> (True, False) -> If -> Bool -> CmpP -> Proj (Result) -> toString;
```
Ah I see the condition above already checks that it can only be a diamond.

eme64 commented on src/hotspot/share/opto/stringopts.cpp:
```diff
@@ -1062,6 +1067,29 @@ bool StringConcat::validate_control_flow() {
+        // where the Bool would be replaced by a constant zero but the Phi stays live
+        // as it is a parameter of the concatenation itself.
+        assert(ptr->in(1)->in(0)->in(1)->is_Bool(), "unexpected if shape");
+        Node* v1 = ptr->in(1)->in(0)->in(1)->in(1)->in(1);
+        Node* v2 = ptr->in(1)->in(0)->in(1)->in(1)->in(2);
```
You may want to use some intermediate results and give them names.
For example:
`Node* iff = ptr->in(1)->in(0)`
You seem to make an assumption that the input of the bool is a cmp, right? Did you check that? Or is it somehow guaranteed? What if in some edge-case of an edge-case it is something else that has only one input? Could that happen?

danielogh commented on test/hotspot/jtreg/compiler/stringopts/TestStackedConcatsPhiUseOfDiamondRegion.java:
```diff
@@ -0,0 +1,57 @@
+        f();
+    }
+
+    static String f() {
+        String s = "a";
+        s = new StringBuilder().append(s).append(s).toString();
+        s = new StringBuilder().append(s).append((s == "xx") ? s : "aa").toString();
+        Asserts.assertEQ(s, "aaaa"); // in particular, we should not have s.equals("aaxx").
+        return s;
+    }
+}
```
I think this is a good idea for sure.

danielogh committed:
direction convention

danielogh committed:
store intermediate calculations

dean-long commented:
Hmm, I see now that validate_control_flow() does limit coalescing, but I'm worried that the pattern matching may not catch all problematic cases.

danielogh commented:
Yes, `validate_control_flow()` is used for individual and coalesced concatenations, but just re-using those same checks for coalesced concatenations has been shown to not be sufficient in recent bugs -- in particular when the result of SB1 is used in unexpected ways in SB2. I am not convinced that we have covered all the cases yet. Would it be an idea to fix this issue and then go for the fuzzing approach next to cover more patterns (follow-up RFE), or is there a more general pattern we could prevent here already?

dean-long commented:
I think we could find a more general pattern, like "no conditional code at all", but that might hurt performance.  It seems like we should be able to handle conditional code and this test case in particular.  I don't have a perfect understanding of this code, but if I was going to try fixing it to handle the problematic cases, I would look at StringConcat::eliminate_unneeded_control(), because it seems to be blindingly removing nodes/edges that can still be needed.

danielogh commented:
A comment to keep this PR active. I think the current approach is at least in the right direction.

As Dean points out, the dependency validation logic might be insufficient (seen in e.g. JDK-8367405), but these diamond patterns are treated as as special case in string concat code. The current bug is the result of assuming diamond-region phis
are not live [1], but that can't be done if they are parameters of the concatenation.

[1] https://github.com/openjdk/jdk/blob/cc5b35bf69dcf9e7e8037642c94e8d7e5847952d/src/hotspot/share/opto/stringopts.cpp#L275

danielogh commented:
/touch

robcasloz reviewed (approved):
The fix looks good to me. I agree we cannot be sure that this changeset discards all invalid cases, but a low-risk point fix in time for JDK 27 is better than waiting indefinitely for a more principled solution, in my opinion. What do you think @eme64 @dean-long?

Having said that, I totally agree that, looking ahead, we should revisit the string concat optimization logic. Its current "denylist" approach makes it hard to be confident that all invalid transformations are rejected.

danielogh commented:
Great, thanks Dean. I'll have a look at the test case you constructed and we should try for a more comprehensive fix.

danielogh committed:
Merge remote-tracking branch 'upstream/master' into stringopts-wrong-result

danielogh committed:
implement allow-list for compare uses and add some tests

danielogh committed:
guard against shared bool case

danielogh commented:
In the latest revision my intention is to only allow compare uses of call results for legitimate null checks. It's implemented by caching valid CmpP uses for append calls in validate_control_flow() and for toString calls in merge() and letting validate_control_flow() check for those uses in its existing result containment analysis. I also added a pre-check for shared bool tests when merging. This revision tests well, does not invalidate any existing optimization, passes the append() case that the earlier revision didn't, as well as all the new stringopts tests @eme64  found such as the one in JDK-8385429. I added those tests to the PR.

Do you think this approach is in the right direction, @dean-long? Thanks.

danielogh commented:
/touch

danielogh commented:
/template append

danielogh commented:
/touch

dean-long commented:
Sorry for the delay.  I somehow missed that this was updated.  I'll take another look.

dean-long commented:
> Do you think this approach is in the right direction, @dean-long? Thanks.

Yes, thanks for not giving up on this one!  I think the current approach looks good.

danielogh commented:
Thanks a lot for reviewing the revised version.

danielogh commented:
/contributor add @eme64

danielogh commented:
Thanks for finding some of these tests

TobiHartmann reviewed (approved):
This looks good to me too. Maybe @robcasloz also wants to re-review. Thanks for working on this!

danielogh commented:
Thanks for the additional approvals!

danielogh commented:
/integrate
